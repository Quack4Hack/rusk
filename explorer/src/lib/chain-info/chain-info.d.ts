type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "block" | "transaction";
};

type Transaction = {
  blockhash: string;
  blockheight: number;
  contract: string;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
};
