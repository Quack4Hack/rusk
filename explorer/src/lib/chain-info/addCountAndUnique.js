/**
 * Returns an array of unique objects, sorted by the frequency of their occurrence,
 * with an additional count property representing the number of times
 * that city-country combination appeared in the input data.
 *
 * @param {Omit<NodeLocation, "count">[]} data
 * @returns {NodeLocation[]}
 */
function addCountAndUnique(data) {
  const countMap = data.reduce((locationMap, location) => {
    const key = `${location.city},${location.country}`;

    if (!locationMap[key]) {
      locationMap[key] = { ...location, count: 0 };
    }
    locationMap[key].count += 1;
    return locationMap;
  }, /** @type {{ [key: string]: NodeLocation }} */ ({}));

  // Return only unique city-country combinations
  return Object.values(countMap).sort((a, b) => b.count - a.count);
}

export default addCountAndUnique;
