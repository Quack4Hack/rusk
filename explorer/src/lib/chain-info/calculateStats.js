import {
  always,
  compose,
  condition,
  filterWith,
  getKey,
  partition,
} from "lamb";

import { arraySumByKey } from "$lib/dusk/array";

/** @type {(txs: Pick<GQLTransaction, "err">[]) => number} */
const getFailedTxAmount = compose(
  getKey("length"),
  filterWith((tx) => tx.err !== null)
);

/**
 * We take into account only provisioners with
 * the minimum stake amount of 1000 Dusk (1e12 Lux).
 *
 * @type {(provisioners: HostProvisioner[]) => HostProvisioner[]}
 */
const getValidProvisioners = filterWith((p) => p.amount >= 1e12);

/**
 * Sums the values of the "amount" key in the received array.
 * Returns zero if the array is empty.
 */
const sumByAmount = condition(
  (provisioners) => provisioners.length > 0,
  arraySumByKey("amount"),
  always(0)
);

/**
 * @param {HostProvisioner[]} provisioners
 * @param {number} lastHeight
 * @param {Pick<GQLTransaction, "err">[]} last100BlocksTxs
 * @returns {Stats}
 */
function calculateStats(provisioners, lastHeight, last100BlocksTxs) {
  const [activeProvisioners, waitingProvisioners] = partition(
    getValidProvisioners(provisioners),
    (p) => p.eligibility <= lastHeight
  );

  return {
    activeProvisioners: activeProvisioners.length,
    activeStake: sumByAmount(activeProvisioners),
    lastBlock: lastHeight,
    txs100blocks: {
      failed: getFailedTxAmount(last100BlocksTxs),
      transfers: last100BlocksTxs.length,
    },
    waitingProvisioners: waitingProvisioners.length,
    waitingStake: sumByAmount(waitingProvisioners),
  };
}

export default calculateStats;
