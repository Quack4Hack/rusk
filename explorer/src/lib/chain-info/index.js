export { default as transformBlock } from "./transformBlock";
export { default as transformTransaction } from "./transformTransaction";
