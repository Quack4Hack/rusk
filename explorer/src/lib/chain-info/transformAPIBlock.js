import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformAPITransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformAPIBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts", "version"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformAPIBlock = ({ header, transactions }) => ({
  header: transformAPIBlockHeader(header),
  transactions: updateIn(
    transactions,
    "data",
    mapWith(transformAPITransaction)
  ),
});

export default transformAPIBlock;
