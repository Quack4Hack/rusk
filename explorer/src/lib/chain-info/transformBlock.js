import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformTransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformBlock = ({ header, transactions }) => ({
  header: transformBlockHeader(header),
  transactions: updateIn(transactions, "data", mapWith(transformTransaction)),
});

export default transformBlock;
