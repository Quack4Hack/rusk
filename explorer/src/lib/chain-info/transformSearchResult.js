import { compose, getPathIn, mapWith } from "lamb";

/**
 * @template {Record<PropertyKey, any>} T
 * @template {string} U
 * @param {T} source
 * @param {U} path
 * @returns {Exclude<import("lamb").GetPath<T, U>, undefined> | []}
 */
const getPathOrEmptyArray = (source, path) => getPathIn(source, path) ?? [];

/** @type {(v: APISearchResult) => SearchResult[]} */
const transformSearchResult = compose(
  mapWith((entry) =>
    entry?.header?.hash
      ? { id: entry.header.hash, type: "block" }
      : { id: entry.txid, type: "transaction" }
  ),
  (v) => [
    ...getPathOrEmptyArray(v, "data.data.blocks"),
    ...getPathOrEmptyArray(v, "data.data.transactions"),
  ]
);

export default transformSearchResult;
