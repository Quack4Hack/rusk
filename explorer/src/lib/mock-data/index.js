export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactionDetails } from "./gql-transaction-details.json";
export { default as gqlTransactions } from "./gql-transactions.json";
