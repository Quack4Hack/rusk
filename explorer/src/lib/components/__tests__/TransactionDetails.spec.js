import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";

import { apiMarketData, gqlTransaction } from "$lib/mock-data";
import { transformTransaction } from "$lib/chain-info";

import { TransactionDetails } from "..";

describe("Transaction Details", () => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(2024, 4, 20));

  const baseProps = {
    data: transformTransaction(gqlTransaction.tx),
    error: null,
    loading: false,
    market: {
      currentPrice: apiMarketData.market_data.current_price,
      marketCap: apiMarketData.market_data.market_cap,
    },
    payload:
      "db0794770322802a22905c4364511f3186e6184085f875dbb9f11a3ae914766c020000000000000014bc23b875c67d0dbecfdd45f5964f3fea7188aff2035730c8802",
  };

  afterEach(cleanup);

  afterAll(() => {
    vi.useRealTimers();
  });

  it("renders the Transaction Details component", () => {
    const { container } = render(TransactionDetails, baseProps);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the Transaction Details component with the payload visible", async () => {
    const { container, getByRole } = render(TransactionDetails, baseProps);

    await fireEvent.click(getByRole("switch"));

    expect(container.firstChild).toMatchSnapshot();
  });
});
