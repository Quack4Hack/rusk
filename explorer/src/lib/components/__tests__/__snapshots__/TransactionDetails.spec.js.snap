// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transaction Details > renders the Transaction Details component 1`] = `
<div
  class="dusk-card data-card transaction-details"
>
  <header
    class="data-card__header"
    slot="header"
  >
    <h1
      class="data-card__header-title"
    >
      Transaction Details
    </h1>
     
  </header>
   
  <div
    class="data-card__content"
  >
    <dl
      class="transaction-details__list"
    >
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The ID of the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        ID
      </dt>
       
      <dd
        class="details-list__definition"
      >
        4877687c2dbf154248d3ddee9ba...39056f82a46819da041d4ac0e04
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The block height this transaction belongs to"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        block height
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <a
          class="dusk-anchor dusk-anchor--on-surface transaction-details__list-anchor"
          href="/some-base-path/blocks/block?id=3c6e4018cfa86723e50644e33d3990bc27fc794f6b49fbf6290e4d308e07bd2d"
        >
          487,166
        </a>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction status"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        Status
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <span
          class="dusk-badge dusk-badge--variant-success transaction-status transaction-details__status explorer-badge"
        >
          success
        </span>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The date and time the transaction was created"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        timestamp
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <time
          class="dusk-relative-time transaction-details__list-timestamp"
          datetime="2024-04-16T06:39:09.000Z"
          slot="definition"
        >
          Tue, 16 Apr 2024 06:39:09 GMT (last month)
          
        </time>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction type"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        type
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transaction-details__type explorer-badge"
        >
          transfer
        </span>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The fee paid for the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        transaction fee
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000290766 DUSK ($0.0001071037)
        
         
        
         
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction gas price"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        gas price
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000000001 DUSK ($0.0000000004)
        
         
        
         
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction gas limit"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        transaction gas limit
      </dt>
       
      <dd
        class="details-list__definition"
      >
        500000000
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The amount of gas spent generating the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        gas spent
      </dt>
       
      <dd
        class="details-list__definition"
      >
        290766
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="Transaction reference and additional notes"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        memo
      </dt>
       
      <dd
        class="details-list__definition"
      >
        some notes about the transaction
        
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The payload"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        payload

        
        <div
          aria-checked="false"
          aria-disabled="false"
          class="dusk-switch transaction-details__payload-switch dusk-switch--on-surface"
          role="switch"
          tabindex="0"
        />
      </dt>
       
      <dd
        class="details-list__definition"
      >
        
      </dd>
    </dl>
  </div>
  
   
</div>
`;

exports[`Transaction Details > renders the Transaction Details component with the payload visible 1`] = `
<div
  class="dusk-card data-card transaction-details"
>
  <header
    class="data-card__header"
    slot="header"
  >
    <h1
      class="data-card__header-title"
    >
      Transaction Details
    </h1>
     
  </header>
   
  <div
    class="data-card__content"
  >
    <dl
      class="transaction-details__list"
    >
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The ID of the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        ID
      </dt>
       
      <dd
        class="details-list__definition"
      >
        4877687c2dbf154248d3ddee9ba...39056f82a46819da041d4ac0e04
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The block height this transaction belongs to"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        block height
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <a
          class="dusk-anchor dusk-anchor--on-surface transaction-details__list-anchor"
          href="/some-base-path/blocks/block?id=3c6e4018cfa86723e50644e33d3990bc27fc794f6b49fbf6290e4d308e07bd2d"
        >
          487,166
        </a>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction status"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        Status
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <span
          class="dusk-badge dusk-badge--variant-success transaction-status transaction-details__status explorer-badge"
        >
          success
        </span>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The date and time the transaction was created"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        timestamp
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <time
          class="dusk-relative-time transaction-details__list-timestamp"
          datetime="2024-04-16T06:39:09.000Z"
          slot="definition"
        >
          Tue, 16 Apr 2024 06:39:09 GMT (last month)
          
        </time>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction type"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        type
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <span
          class="dusk-badge dusk-badge--variant-neutral transaction-details__type explorer-badge"
        >
          transfer
        </span>
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The fee paid for the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        transaction fee
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000290766 DUSK ($0.0001071037)
        
         
        
         
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction gas price"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        gas price
      </dt>
       
      <dd
        class="details-list__definition"
      >
        0.000000001 DUSK ($0.0000000004)
        
         
        
         
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The transaction gas limit"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        transaction gas limit
      </dt>
       
      <dd
        class="details-list__definition"
      >
        500000000
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The amount of gas spent generating the transaction"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        gas spent
      </dt>
       
      <dd
        class="details-list__definition"
      >
        290766
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="Transaction reference and additional notes"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        memo
      </dt>
       
      <dd
        class="details-list__definition"
      >
        some notes about the transaction
        
      </dd>
      <dt
        class="details-list__term"
      >
        <svg
          class="dusk-icon dusk-icon--size--normal detail-list__help"
          data-tooltip-id="main-tooltip"
          data-tooltip-place="top"
          data-tooltip-text="The payload"
          data-tooltip-type="info"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"
          />
        </svg>
        
         
        payload

        
        <div
          aria-checked="true"
          aria-disabled="false"
          class="dusk-switch transaction-details__payload-switch dusk-switch--on-surface"
          role="switch"
          tabindex="0"
        />
      </dt>
       
      <dd
        class="details-list__definition"
      >
        <div
          class="dusk-card transaction-details__payload dusk-card--on-surface"
        >
           
          db0794770322802a22905c4364511f3186e6184085f875dbb9f11a3ae914766c020000000000000014bc23b875c67d0dbecfdd45f5964f3fea7188aff2035730c8802
          
           
        </div>
        
        
      </dd>
    </dl>
  </div>
  
   
</div>
`;
